### ISSUE TITLE
Private Cluster V2

### ISSUE THREAD

**Issue by miwithro** [Azure/AKS#2729]
Private Cluster is currently based on Private Link, and V2 will move to SWIFT.

**Comment by rouke-broersma** [comment]
@miwithro could you add some documentation on what SWIFT is? We use private link for private clusters with user created private endpoints so we can connect to the cluster API from different tenants so it sounds like this could have a large impact on our workflows.

**Comment by hbuckle** [comment]
Is it the same Swift Virtual Network Connection that app service uses? We haven't found that to be particularly reliable.

**Comment by carvido1** [comment]
> @miwithro could you add some documentation on what SWIFT is? We use private link for private clusters with user created private endpoints so we can connect to the cluster API from different tenants so it sounds like this could have a large impact on our workflows.

same as @miwithro , it would be nice to have clarifications from microsoft, as we are working on recreating AKS clusters using AKS private clusters. Maybe this is related ? https://www.swift.com/our-solutions/interfaces-and-integration/public-cloud

**Comment by miwithro** [comment]
@rouke-broersma @hbuckle @carvido1 API Server VNET Integration allows us from a topology perspective to get rid of the tunnel-front, openvpn, konnectivity pods for access.  So it is a more secure by default network topology that will NOT ONLY apply to Private AKS but also Public AKS.

**Comment by rouke-broersma** [comment]
@miwithro will vnet integration replace private link or will it be an alternative to it? We currently use private link with aks to enable connectivity from non-peered vnets (multiple private endpoint connections to the same aks). Would we be able to continue using this or will we need to create an alternative approach?

**Comment by miwithro** [comment]
No, vnet integration will replace Private Link between the AKS Nodes and the Control Plane, your existing connectivity from non-peered vnets will continue as is.

**Comment by hbuckle** [comment]
So to confirm, we'll be able to create multiple vnet integrations in different vnets?
Will there be a requirement for dedicated subnets like with app service?
Will we still need to create private DNS zones like with private link?

**Comment by rouke-broersma** [comment]
@hbuckle how I understand it is that only the control plane will use vnet integration for the communication between the control plane and the aks nodes. Communication from clients to the control plane will still use private link. Is that correct @miwithro ?

**Comment by miwithro** [comment]
@rouke-broersma that is correct.

**Comment by phealy** [comment]
This feature was released to public preview on 2022 June 27. [Documentation is available in the standard AKS documentation site.](https://docs.microsoft.com/en-us/azure/aks/api-server-vnet-integration)

@hbuckle one cluster will only be VNet integrated into a single VNet. You'll be able to create private link services on the kube-apiserver ILB to expose it across virtual networks, though that functionality doesn't *yet* work at public preview. Yes, the API server subnet must be delegated to Microsoft.ContainerService/managedClusters, but you can use it for multiple clusters as long as it's big enough. Private DNS zones will still be created by default, but the nodes and cluster workloads reach the API VIP directly by IP (since it's in a delegated subnet, the IP can't be changed, so we don't need to use DNS). You can disable provisioning of the private DNS zone and just use the public FQDN record we create by following the regular [private cluster DNS setup steps](https://docs.microsoft.com/azure/aks/private-clusters#configure-private-dns-zone).

@rouke-broersma VNet integration will be used for all communication between your private network, including nodes and out-of-cluster kubernetes clients, and the API server. Private link will only be used if you need to expose the cluster to a separate VNet without peering - e.g. you deploy AKS in an "island" unpeered virtual network, then you create a PLS pointing at kube-apiserver and create a PE to expose the API server endpoint into your main hub/spoke for connectivity. That functionality is planned but PE/PLS support is not available as of initial public preview. Importantly, it will be self-managed - AKS with API server VNet integration won't create the PLS for you, so you have complete control over how it's provisioned.

**Comment by rouke-broersma** [comment]
@phealy so if I understand you correctly the previous information provided by @miwithro was incomplete?

I had asked before if vnet integration would replace the current private link setup and was told that it would. I assume that my question was misunderstood as meaning if you enable it, it will replace private link.
What I meant was, will vnet integration be an option for private clusters (ie private link or vnet integration), or will we be forced to migrate to vnet integration for private clusters?

**Comment by kevinharing** [comment]
With this change, will it be possible to whitelist the API server subnet on a storage account firewall, so that the csi-azurefile-controller can communicate through the data plane instead of the control plane of the storage account? Or are the managed controllers hosted elsewhere?

**Comment by PixelRobots** [comment]
@phealy Any update on when new regions will be supported in the preview?

**Comment by miwithro** [comment]
I just updated the document yesterday.  We now support ALL Public regions minus "southcentralus"

**Comment by paulgmiller** [comment]
What ports are used are actually work load dependent. web hooks, kubectl port-forward ,  proxy can all use different ports. kubectl exec, logs probably have standard ports on kublet.

**Comment by zhiweiv** [comment]
Just want to confirm again, Any ETA of GA?

**Comment by PixelRobots** [comment]
@miwithro or @phealy Do you have an estimate for GA?

**Comment by cwe1ss** [comment]
comment to keep it active. any update about the ETA for GA would still be great.

**Comment by levimm** [comment]
We don't have exact GA date yet. For those who are deciding whether to create tradition private cluster or wait for apiserver vnet integration GA, they can just go and create tradition private cluster for now. We would support upgrade from tradition cluster (no matter private or not) to apiserver vnet integration.

**Comment by phealy** [comment]
> @rouke-broersma @hbuckle @carvido1 API Server VNET Integration allows us from a topology perspective to get rid of the tunnel-front, openvpn, konnectivity pods for access.

Note for those reading this thread: this was correct at the time of writing, but Konnectivity will still be deployed for API Server VNet Integration clusters that utilize Azure CNI Overlay or BYOCNI (new features since that comment was written). In these scenarios communication between the API server and the nodes will still be direct, but the API server can't reach the pods directly because they're in an overlay network. Konnectivity will be used for traffic from the API server to destinations in the PodCIDR - typically for connectivity to metrics-server and for supporting `kubectl port-forward`, but it will also be used if any webhook configurations are created that are hosted in-cluster.

**Comment by kevinharing** [comment]
Any update on when this will be GA? It’s been a while without any updates.

**Comment by PixelRobots** [comment]
Hey @chasewilson any update on this going GA?

**Comment by chasewilson** [comment]
Unfortunately no updates yet. We're still working on the feature but have run into some dependencies we are working to overcome.

**Comment by PixelRobots** [comment]
Hey @chasewilson any further updates on this issue?

**Comment by erezbatish-ms** [comment]
Any updates on GA date?

**Comment by kevinharing** [comment]
Really want to use this in production, any update on when it will become GA?

**Comment by rafallopatka** [comment]
Any updates on GA date?

**Comment by miwithro** [comment]
We are trying to GA in Q4 2024.  We are working with closely with Azure Networking to ensure the dependencies across Azure regions are in place to support API Server VNET integration.  As we get closer to the date, we will update the github link for everyone to be aware.

**Comment by simon-harvey-evelyn** [comment]
@miwithro is this still planned for Q4 2024?

**Comment by chasewilson** [comment]
> @miwithro is this still planned for Q4 2024?

This is still planned for this year and we're closely watching this one with high priority.

**Comment by illrill** [comment]
Need guidance on how large the subnet should be. The docs say we need minimum `/28` to _function_, but what do we need to _scale_? For example, some recommendations for clusters of 500 / 1000 / 2000 / 5000 nodes?

**Comment by paolo-rendano** [comment]
@miwithro we are planning to go on production with this feature enabled by the end of September. Is there any update on the ETA?
Once the feature will go GA shall the cluster need to be recreated?
Thanks

Paolo

**Comment by PixelRobots** [comment]
> @miwithro @chasewilson we are planning to go on production with this feature enabled by the end of September. Is there any update on the ETA? Once the feature will go GA shall the cluster need to be recreated? Thanks
> 
> Paolo

GA daste has been updated in the first comment:

![image](https://github.com/user-attachments/assets/aaeb2d6c-0803-4391-a249-43701c07d19b)

**Comment by jemag** [comment]
regarding the pinned comment @miwithro , could you also confirm if the transition will also be seamless for users for the current preview feature once it becomes GA?

**Comment by Caroline132** [comment]
We are also about to deploy several clusters and would love to have confirmation that our "preview" VNET integrated clusters won't need to be recreated once the features goes GA.

**Comment by cmeury** [comment]
We are in the same situation, working towards production and wondering about seamless transition. Also, is the ETA of Q1 2025 still valid?

**Comment by chasewilson** [comment]
No current updates. However, I want to reiterate that the ETA is an estimation. While I wish i could provide that date with confidence, I want to make sure expectations are set accordingly. This is something we're watching closely and any update, slip or otherwise, we'll update here immediately.

**Comment by jkroepke** [comment]
Would it be possible to move this feature into the stable API version, without promoting to GA? Then, the flag can be integrated in Terraform again and most of the users should be happy again.

**Comment by rowanmoul** [comment]
While I would personally be on board with that since I am stuck on Terraform/Azurerm 3.x until this drops, I think that defeats the point of calling an api "stable".

**Comment by paolo-rendano** [comment]
Same here, we are also stuck on terraform 3 with several environments

**Comment by tspearconquest** [comment]
Hi team, we REALLY need an update on when this will be GA.

We CANNOT upgrade to provider v4 until this feature is GA due to Microsoft forcing Hashicorp to remove support for preview features in the API once they released v4, because doing so would expose our API server to the public internet, instead of creating a load balancer in a private subnet.

Because of the inability to upgrade to provider v4, we're receiving notifications in the portal now that Istio support via ASM 1.22 is ending on March 10 and that it will have no support unless we upgrade to ASM 1.23.

We tried to upgrade to the v4 provider, however because we are on azurerm v3.x, where the `azurerm_kubernetes_cluster` block `service_mesh_profile` does not support the `revisions` tag, we are unable to tell the Azure API to install ASM 1.23 or ASM 1.24, and so we are left unable to upgrade either the terraform provider or the ASM release, and will be unable to receive support post to March 10.

**Comment by chasewilson** [comment]
I regret to inform everyone here that this is slipping just a bit again. 

However, it's slipping "less", if I could put it that way. We are working on getting it to GA early Q2 25 (CY not FY) but the approach we've had to take to get it there is taking some extra planning. Thank you everyone for your patience here we're not letting up on this feature at all and will keep pushing as hard as we can!

**Comment by jkroepke** [comment]
@chasewilson To be honest, not many people care about GA or Preview itself. The real issue is that Microsoft is pushing HashiCorp to exclusively use the GA API for AKS. That’s why so many people are asking for GA status. However, you could help a lot of users by allowing HashiCorp to use the AKS preview API again.

I’m also unsure why this restriction applies only to AKS, while other resources can still take advantage of preview features.

After all, using Bicep lets end-users enable preview features too — so this feels more like a soft limitation.

**Comment by tspearconquest** [comment]
@chasewilson thank you for the update. However, as paying enterprise customers, I think I speak for many of us here and many more who are unaware of this GH issue when I say this delay is unacceptable. I will be opening a support case and raising this issue internally within my org to the highest levels I can get it to.

The fact that this feature was not prioritized for GA for 2+ years and then ripped away from the largest tool people use is mind boggling.

@jkroepke is 100% correct too. Microsoft is not going to get people to switch to bicep; my org has adopted a company wide strategy leveraging terraform and so switching to Bicep for one specific resource is never going to happen. If anything, I'm going to have to switch to the AzAPI provider to get us into a state where we can upgrade Istio, which is a major rework to workaround a restriction that's been imposed on IBM/Hashicorp at effectively the last minute. If Microsoft deprecates any of the API calls that the v3.x azurerm provider uses between now and whenever this does reach GA, there is going to be a very large risk of real harm to Microsoft's own customers due to being unable to take the upgrade. 

@stephybun FYSA

**Comment by rowanmoul** [comment]
> but the approach we've had to take to get it there is taking some extra planning.

While this delay is disappointing, I suspect that many people don't fully appreciate the scale that Azure operates on (I certainly struggle to). It takes time and planning to execute a seamless infrastructure change even at my comparatively microscopic company. I can only imagine the amount of planning that has to go into rolling out new features to Azure data centres all around the world without disrupting the millions of businesses that rely on Azure to operate every day.
I'd rather this feature have another month to cook than roll out and break my production workloads that are already using the preview version.

**Comment by tspearconquest** [comment]
Definitely no disagreement on that! I'm not intending to imply that they should rush this; in fact I wish they would slow down the deprecations. That's the concern for me and I realize now I didn't make that clear before.

**Comment by dixneuf19** [comment]
Hi,
Also a paying customer of Azure, we would be interested to have this feature GA to less network errors in our AKS cluster when contacting the API server. 🙏

**Comment by abubinski** [comment]
Updating the GA ETA to May 19, 2025.

**Comment by PixelRobots** [comment]
Just in time for Build!

**Comment by stijndehaes** [comment]
I believe this is supposed to go GA today, will you post an update here or is there another place where we can follow up if it is released and which regions are supported? :)

**Comment by tspearconquest** [comment]
> is there another place where we can follow up if it is released and which regions are supported? :)

I'm not a Microsoft employee so I can't say for sure but they generally announce these things on the release tracker: https://github.com/Azure/AKS/releases

**Comment by thibfilipe** [comment]
@abubinski still no news on the GA 3 days after the ETA? 🙄

**Comment by PixelRobots** [comment]
Looks like it is GA. https://aka.ms/aks/new

https://learn.microsoft.com/en-us/azure/aks/api-server-vnet-integration#limited-availability

**Comment by jkroepke** [comment]
GA is GA, the terraform integration can be started.

**Comment by tspearconquest** [comment]
I agree, this isn't **exactly** GA - very limited capacity, only a handful of regions, and a warning about stopping the cluster deallocating the control plane and releasing that capacity.

Question: if I'm still using an vnet integration preview cluster in a region where the GA version is not available, does stopping the cluster in that scenario deallocate the control plane and if yes, would I potentially run into the same issue?

**Comment by rowanmoul** [comment]
It looks like it's been in the "stable API" since March (API version 2025-03-01), so I think as far as Terraform is concerned, this can be added back in.
Compare [March](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?view=rest-aks-2025-03-01&tabs=HTTP#managedclusterapiserveraccessprofile) with [February](https://learn.microsoft.com/en-us/rest/api/aks/managed-clusters/create-or-update?view=rest-aks-2025-02-01&tabs=HTTP#managedclusterapiserveraccessprofile)

That said, I'd like to hear more about what if anything those of us using the preview might have to do to "migrate". The rest api property is the same, so I presume we don't have to do anything? (compare the above with [this preview api spec](https://github.com/Azure/azure-rest-api-specs/blob/803d936c40c409ca5bb0106f3155ddb50fdb5408/specification/containerservice/resource-manager/Microsoft.ContainerService/aks/preview/2022-04-02-preview/managedClusters.json#L4498))

**Comment by PixelRobots** [comment]
@chasewilson are you able to add anything to this?

**Comment by thpham** [comment]
Do we have an estimated ETA for GA in more regions ? like Switzerland North and Germany West Central ? Where could we find such informations ?! thank you.

**Comment by tspearconquest** [comment]
Add EastUS also please.

**Comment by cjedro** [comment]
@thpham @tspearconquest both are now available: https://learn.microsoft.com/en-us/azure/aks/api-server-vnet-integration#limited-availability

**Comment by chrisjohnson00** [comment]
@abubinski  @chasewilson 

> API Server VNet Integration is available in all GA public cloud regions except eastus2 and qatarcentral.

Any update/etc on full region support?
